Description: Check whether xcb-screensaver extension is present
Origin: https://github.com/telegramdesktop/tdesktop/commit/6635d03818c90f62cfddb516acadc293255b8905
Bug: https://github.com/telegramdesktop/tdesktop/issues/8506
Bug-Debian: https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=976894
From: Ilya Fedin <fedin-ilja2010@ya.ru>
Date: Sat, 29 Aug 2020 03:55:38 +0400
Acked-By: Nicholas Guriev <guriev-ns@ya.ru>
Last-Update: Wed, 16 Dec 2020 17:56:27 +0300

--- a/Telegram/SourceFiles/platform/linux/specific_linux.cpp
+++ b/Telegram/SourceFiles/platform/linux/specific_linux.cpp
@@ -324,6 +324,20 @@ bool GenerateDesktopFile(
 	}
 }
 
+bool IsXCBExtensionPresent(
+		xcb_connection_t *connection,
+		xcb_extension_t *ext) {
+	const auto reply = xcb_get_extension_data(
+		connection,
+		ext);
+
+	if (!reply) {
+		return false;
+	}
+
+	return reply->present;
+}
+
 std::optional<crl::time> XCBLastUserInputTime() {
 	if (const auto native = QGuiApplication::platformNativeInterface()) {
 		const auto connection = reinterpret_cast<xcb_connection_t*>(
@@ -333,6 +347,10 @@ std::optional<crl::time> XCBLastUserInpu
 			return std::nullopt;
 		}
 
+		if (!IsXCBExtensionPresent(connection, &xcb_screensaver_id)) {
+			return std::nullopt;
+		}
+
 		const auto screen = xcb_setup_roots_iterator(
 			xcb_get_setup(connection)).data;
 
