Description: Workaround for ICE in requestFileReference template
Author: Nicholas Guriev <guriev-ns@ya.ru>
Bug: https://github.com/telegramdesktop/tdesktop/issues/5111
Forwarded: https://github.com/telegramdesktop/tdesktop/pull/5124
Last-Update: Thu, 30 Aug 2018 23:17:41 +0300

--- a/Telegram/SourceFiles/apiwrap.cpp
+++ b/Telegram/SourceFiles/apiwrap.cpp
@@ -2390,7 +2390,11 @@ void ApiWrap::requestFileReference(
 
 	request(std::move(data)).done([=](const auto &result) {
 		const auto parsed = Data::GetFileReferences(result);
-		for (const auto &[origin, reference] : parsed) {
+		for (const auto &p : parsed) {
+			// Unpack here the parsed pair by hand to workaround a GCC bug.
+			// See https://gcc.gnu.org/bugzilla/show_bug.cgi?id=87122
+			const auto &origin = p.first;
+			const auto &reference = p.second;
 			const auto documentId = base::get_if<DocumentFileLocationId>(
 				&origin);
 			if (documentId) {
