Description: Include a missing header in the codegen subproject
Forwarded: https://github.com/desktop-app/codegen/pull/11
Author: Nicholas Guriev <guriev-ns@ya.ru>
Last-Update: Thu, 25 Sep 2021 18:13:01 +0300

--- a/Telegram/codegen/codegen/emoji/options.h
+++ b/Telegram/codegen/codegen/emoji/options.h
@@ -6,8 +6,8 @@
 //
 #pragma once
 
+#include <vector>
 #include <QtCore/QString>
-#include <QtCore/QStringList>
 
 namespace codegen {
 namespace emoji {
