From: Nicholas Guriev <guriev-ns@ya.ru>
Date: Tue, 22 Aug 2017 16:15:36 +0300
Subject: [PATCH] Packed resources
Bug: https://github.com/telegramdesktop/tdesktop/pull/3817
Last-Update: Tue, 10 Sep 2019 21:37:13 +0300

--- a/Telegram/SourceFiles/core/file_utilities.cpp
+++ b/Telegram/SourceFiles/core/file_utilities.cpp
@@ -16,6 +16,7 @@ https://github.com/telegramdesktop/tdesk
 
 #include <QtWidgets/QFileDialog>
 #include <QtCore/QCoreApplication>
+#include <QtCore/QResource>
 #include <QtCore/QStandardPaths>
 #include <QtGui/QDesktopServices>
 
@@ -398,3 +399,20 @@ bool GetDefault(
 
 } // namespace internal
 } // namespace FileDialog
+
+void Resources::LoadAllData() {
+#ifdef TDESKTOP_USE_PACKED_RESOURCES
+	// Load resources packed into a separated file
+	QStringList paths;
+#ifdef _DEBUG
+	paths += cExeDir();
+#endif
+	paths += QStandardPaths::standardLocations(QStandardPaths::AppDataLocation);
+	for (QString directory : paths) {
+		if (QResource::registerResource(directory + qsl("/tresources.rcc"))) {
+			return;  // found
+		}
+	}
+	qFatal("Packed resources not found");
+#endif
+}
--- a/Telegram/SourceFiles/core/file_utilities.h
+++ b/Telegram/SourceFiles/core/file_utilities.h
@@ -107,3 +107,9 @@ bool GetDefault(
 
 } // namespace internal
 } // namespace FileDialog
+
+namespace Resources {
+
+void LoadAllData();
+
+} // namespace Resources
--- a/Telegram/SourceFiles/core/launcher.cpp
+++ b/Telegram/SourceFiles/core/launcher.cpp
@@ -12,6 +12,7 @@ https://github.com/telegramdesktop/tdesk
 #include "platform/platform_info.h"
 #include "ui/main_queue_processor.h"
 #include "core/crash_reports.h"
+#include "core/file_utilities.h"
 #include "core/update_checker.h"
 #include "core/sandbox.h"
 #include "base/concurrent_timer.h"
@@ -244,6 +245,7 @@ void Launcher::init() {
 	prepareSettings();
 
 	QApplication::setApplicationName(qsl("TelegramDesktop"));
+	Resources::LoadAllData(); // should be called after setting an application name
 
 #ifdef TDESKTOP_LAUNCHER_FILENAME
 #define TDESKTOP_LAUNCHER_FILENAME_TO_STRING_HELPER(V) #V
--- a/Telegram/gyp/telegram/telegram.gypi
+++ b/Telegram/gyp/telegram/telegram.gypi
@@ -48,6 +48,7 @@
       ],
       'build_defines%': '',
       'use_common_xxhash%': 0,
+      'use_packed_resources%': 0,
       'list_sources_command': 'python <(DEPTH)/list_sources.py --input <(DEPTH)/telegram/sources.txt --replace src_loc=<(src_loc)',
       'pch_source': '<(src_loc)/stdafx.cpp',
       'pch_header': '<(src_loc)/stdafx.h',
@@ -112,7 +113,6 @@
       '/usr/include/libtgvoip',
     ],
     'sources': [
-      '<@(qrc_files)',
       '<@(style_files)',
       '<(res_loc)/langs/cloud_lang.strings',
       '<(res_loc)/export_html/css/style.css',
@@ -126,6 +126,31 @@
       '<!@(<(list_sources_command) <(qt_moc_list_sources_arg) --exclude_for <(build_os))',
     ],
     'conditions': [
+      [ 'use_packed_resources', {
+        'actions': [
+          {
+            'action_name': 'generate_resource_pack',
+            'inputs': [
+              '<(SHARED_INTERMEDIATE_DIR)/update_dependent_qrc.timestamp',
+            ],
+            'outputs': [
+              '<(PRODUCT_DIR)/tresources.rcc',
+            ],
+            'action': [
+              '<(qt_bindir)/rcc<(exe_ext)', '-binary',
+              '<@(qrc_files)',
+              '-o', '<@(_outputs)',
+            ],
+          },
+        ],
+        'defines': [
+          'TDESKTOP_USE_PACKED_RESOURCES',
+        ],
+      }, {
+        'sources': [
+          '<@(qrc_files)',
+        ],
+      }],
       [ '"<(official_build_target)" != ""', {
         'defines': [
           'TDESKTOP_OFFICIAL_TARGET=<(official_build_target)',
