Description: Open Sans Semibold instead of Open Sans Bold
Author: Nicholas Guriev <guriev-ns@ya.ru>
Date: Sun, 10 May 2020 00:14:20 +0300

--- a/Telegram/lib_ui/ui/style/style_core_font.cpp
+++ b/Telegram/lib_ui/ui/style/style_core_font.cpp
@@ -318,7 +318,7 @@ FontData::FontData(int size, uint32 flag
 	}
 
 	f.setPixelSize(size);
-	f.setBold((_flags & FontBold) || (_flags & FontSemibold));
+	f.setBold(_flags & FontBold);
 	f.setItalic(_flags & FontItalic);
 	f.setUnderline(_flags & FontUnderline);
 	f.setStrikeOut(_flags & FontStrikeOut);
