Description: Set font style separately for bold and semibold
Bug-Debian: https://bugs.debian.org/961526
Author: Nicholas Guriev <guriev-ns@ya.ru>
Date: Fri, 29 May 2020 23:29:54 +0300

--- a/Telegram/lib_ui/ui/style/style_core_font.cpp
+++ b/Telegram/lib_ui/ui/style/style_core_font.cpp
@@ -316,14 +316,11 @@ FontData::FontData(int size, uint32 flag
 	f.setItalic(_flags & FontItalic);
 	f.setUnderline(_flags & FontUnderline);
 	f.setStrikeOut(_flags & FontStrikeOut);
-
-	if ((_flags & FontBold) || (_flags & FontSemibold)) {
-#ifdef DESKTOP_APP_USE_PACKAGED_FONTS
-		f.setWeight(QFont::DemiBold);
-#else // DESKTOP_APP_USE_PACKAGED_FONTS
-		f.setBold(true);
-		f.setStyleName("Semibold");
-#endif // !DESKTOP_APP_USE_PACKAGED_FONTS
+	f.setBold(_flags & FontBold);
+	if (_flags & FontItalic) {
+		f.setStyleName(_flags & FontSemibold ? "Semibold Italic" : "Italic");
+	} else {
+		f.setStyleName(_flags & FontSemibold ? "Semibold" : "Regular");
 	}
 
 	m = QFontMetrics(f);
